### PR TITLE
[SYCL] Enable SPV_INTEL_fpga_invocation_pipelining_attributes extension

### PR DESCRIPTION
**Pull request by MrSidims**
Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -8611,7 +8611,8 @@ void SPIRVTranslator::ConstructJob(Compilation &C, const JobAction &JA,
         ",+SPV_INTEL_arbitrary_precision_floating_point"
         ",+SPV_INTEL_variable_length_array,+SPV_INTEL_fp_fast_math_mode"
         ",+SPV_INTEL_fpga_cluster_attributes,+SPV_INTEL_loop_fuse"
```
Comment (MrSidims) [inline clang/lib/Driver/ToolChains/Clang.cpp:8613]: There is actually a gap in our design of how we enable SPIR-V extensions. I'm 100%, that Intel GPU doesn't support most of Intel FPGA extensions, but that issue is going to be addressed later.

File: clang/test/Driver/sycl-spirv-ext.c [hunk 2]
```diff
@@ -25,5 +25,5 @@
...
+// CHECK-FPGA-HW: llvm-spirv{{.*}}"-spirv-ext=-all,+SPV_EXT_shader_atomic_float_add,+SPV_EXT_shader_atomic_float_min_max,+SPV_KHR_no_integer_wrap_decoration,+SPV_KHR_float_controls,+SPV_INTEL_subgroups,+SPV_INTEL_media_block_io,+SPV_INTEL_device_side_avc_motion_estimation,+SPV_INTEL_fpga_loop_controls,+SPV_INTEL_fpga_memory_attributes,+SPV_INTEL_fpga_memory_accesses,+SPV_INTEL_unstructured_loop_controls,+SPV_INTEL_fpga_reg,+SPV_INTEL_blocking_pipes,+SPV_INTEL_function_pointers,+SPV_INTEL_kernel_attributes,+SPV_INTEL_io_pipes,+SPV_INTEL_inline_assembly,+SPV_INTEL_arbitrary_precision_integers,+SPV_INTEL_optimization_hints,+SPV_INTEL_float_controls2,+SPV_INTEL_vector_compute,+SPV_INTEL_fast_composite,+SPV_INTEL_fpga_buffer_location,+SPV_INTEL_arbitrary_precision_fixed_point,+SPV_INTEL_arbitrary_precision_floating_point,+SPV_INTEL_arbitrary_precision_floating_point,+SPV_INTEL_variable_length_array,+SPV_INTEL_fp_fast_math_mode,+SPV_INTEL_fpga_cluster_attributes,+SPV_INTEL_loop_fuse,+SPV_INTEL_long_constant_composite,+SPV_INTEL_usm_storage_classes,+SPV_INTEL_invocation_pipelining_attributes"
```
Comment (AGindinson) [inline clang/test/Driver/sycl-spirv-ext.c:29]: As a side-note, I wonder if using `CHECK-SAME` would make the list easier to parse and to rebase across different versions of the file.

**Review by AGindinson — APPROVED** [review]

File: clang/test/Driver/sycl-spirv-ext.c (hunk 2 above)
Comment (MrSidims) [reply clang/test/Driver/sycl-spirv-ext.c:29]: You are right, let me apply it in this PR.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -8611,7 +8611,8 @@ void SPIRVTranslator::ConstructJob(Compilation &C, const JobAction &JA,
...
-        ",+SPV_INTEL_long_constant_composite";
+        ",+SPV_INTEL_long_constant_composite"
+        ",+SPV_INTEL_invocation_pipelining_attributes";
```
Comment (MrSidims) [inline clang/lib/Driver/ToolChains/Clang.cpp:8615]: incorrect name

File: clang/test/Driver/sycl-spirv-ext.c [hunk 4]
```diff
@@ -25,5 +25,46 @@
...
-// CHECK-FPGA-HW: llvm-spirv{{.*}}"-spirv-ext=-all,+SPV_EXT_shader_atomic_float_add,+SPV_EXT_shader_atomic_float_min_max,+SPV_KHR_no_integer_wrap_decoration,+SPV_KHR_float_controls,+SPV_INTEL_subgroups,+SPV_INTEL_media_block_io,+SPV_INTEL_device_side_avc_motion_estimation,+SPV_INTEL_fpga_loop_controls,+SPV_INTEL_fpga_memory_attributes,+SPV_INTEL_fpga_memory_accesses,+SPV_INTEL_unstructured_loop_controls,+SPV_INTEL_fpga_reg,+SPV_INTEL_blocking_pipes,+SPV_INTEL_function_pointers,+SPV_INTEL_kernel_attributes,+SPV_INTEL_io_pipes,+SPV_INTEL_inline_assembly,+SPV_INTEL_arbitrary_precision_integers,+SPV_INTEL_optimization_hints,+SPV_INTEL_float_controls2,+SPV_INTEL_vector_compute,+SPV_INTEL_fast_composite,+SPV_INTEL_fpga_buffer_location,+SPV_INTEL_arbitrary_precision_fixed_point,+SPV_INTEL_arbitrary_precision_floating_point,+SPV_INTEL_arbitrary_precision_floating_point,+SPV_INTEL_variable_length_array,+SPV_INTEL_fp_fast_math_mode,+SPV_INTEL_fpga_cluster_attributes,+SPV_INTEL_loop_fuse,+SPV_INTEL_long_constant_composite,+SPV_INTEL_usm_storage_classes"
+// CHECK-DEFAULT: llvm-spirv{{.*}}"-spirv-ext=-all,+SPV_EXT_shader_atomic_float_add
```
Comment (AGindinson) [inline clang/test/Driver/sycl-spirv-ext.c:28]:
```suggestion
// CHECK-DEFAULT: llvm-spirv{{.*}}"-spirv-ext=-all
// CHECK-DEFAULT-SAME:,+SPV_EXT_shader_atomic_float_add
```

File: clang/test/Driver/sycl-spirv-ext.c [hunk 5]
```diff
@@ -25,5 +25,46 @@
...
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_vector_compute,+SPV_INTEL_fast_composite
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_fpga_buffer_location
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_arbitrary_precision_fixed_point
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_arbitrary_precision_floating_point
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_arbitrary_precision_floating_point
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_variable_length_array,+SPV_INTEL_fp_fast_math_mode
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_fpga_cluster_attributes,+SPV_INTEL_loop_fuse
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_long_constant_composite
+// CHECK-DEFAULT-SAME:,+SPV_INTEL_invocation_pipelining_attributes"
+// CHECK-FPGA-HW: llvm-spirv{{.*}}"-spirv-ext=-all,+SPV_EXT_shader_atomic_float_add
```
Comment (AGindinson) [inline clang/test/Driver/sycl-spirv-ext.c:49]:
```suggestion
// CHECK-FPGA-HW: llvm-spirv{{.*}}"-spirv-ext=-all
// CHECK-FPGA-HW-SAME:,+SPV_EXT_shader_atomic_float_add
```

File: clang/test/Driver/sycl-spirv-ext.c [hunk 6]
```diff
@@ -25,5 +25,46 @@
...
+// CHECK-DEFAULT: llvm-spirv{{.*}}"-spirv-ext=-all,+SPV_EXT_shader_atomic_float_add
+// CHECK-DEFAULT-SAME:,+SPV_EXT_shader_atomic_float_min_max
```
Comment (mlychkov) [inline clang/test/Driver/sycl-spirv-ext.c:29]: Not for current review, but a thought for a future: may we add some alias for grouping extension options? If someday there will be 100+ supported extensions then llvm-spirv command string will be very large.

Comment (MrSidims) [reply clang/test/Driver/sycl-spirv-ext.c:29]: Such aliases would be controlled by the Khronos translator, and it now knows +all and -all. We just moved from +all to something more explicit to have control over extensions, that changes translation flow for common LLVM IR patterns, so devices' backend could adapt first.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply clang/lib/Driver/ToolChains/Clang.cpp:8613]: If SPIR-V extension can only be triggered by corresponding DPC++ feature and we are clear in our documentation/intent that it is not supported on GPU, it should be fine

**Review by AGindinson — DISMISSED** [review]

File: clang/test/Driver/sycl-spirv-ext.c (hunk 6 above)
Comment (AlexeySachkov) [reply clang/test/Driver/sycl-spirv-ext.c:30]: I think that some wildcards could be introduced to simplify the string, something like: `+SPV_INTEL_FPGA_*` to enable all FPGA-related extensions. However, such wildcard won't help us and most likely have the opposite effect: some FPGA-specific extensions which are not yet supported by BEs will be enabled as soon as they known to the translator.

Having some hardcoded groups would be a better approach probably, but again: FPGA-specific group can be extended as soon as extension is known to the translator (which seems quite logical), but we might want to postpone that until the extension is adopted by BEs.

The idea seems helpful in general, but I don't have any particular suggestions about how to make it useful for our particular use-case

**Comment by MrSidims** [comment]
From the test failure:

> Unknown extension 'SPV_INTEL_fpga_invocation_pipelining_attributes'

Indeed we need a pull-down first.

**Comment by bader** [comment]
I converted this pull request to a draft to avoid unintentional merge.

**Comment by keryell** [comment]
I am curious about what does it means for an FPGA attribute not to be supported on GPU or CPU or...
What happened if they are just ignored, like any other attribute?
At the end it would be nice to have a SYCL program just to work on any target and have  functionality portability.

**Comment by AGindinson** [comment]
> I am curious about what it means for an FPGA attribute not to be supported on GPU or CPU or...
> What happens if they are just ignored, like any other attribute?
> At the end it would be nice to have a SYCL program just to work on any target and have functionality portability.

@keryell, if I understand correctly, the reference here is made to [this comment thread](https://github.com/intel/llvm/pull/3864#discussion_r643807550) about the fact that we enable a whole set of SPIR-V extensions for any target regardless of whether these extensions are supported by the underlying device runtimes. In this light, "ignoring" unknown SPIR-V capabilities is simply not how the device runtimes are currently taught to operate - and changing their behavior is not something we could do on the SYCL level. If LLVM BC and not SPIR-V was used as the interface between the FE and device backends, metadata-based attributes (i.e. most of the attributes) could truly be ignored by "irrelevant" runtimes. But the latter is an unrealistic architectural change with too many drawbacks.

We could, however, search for ways to disable FPGA-specific extensions for non-FPGA targets within the compiler driver, thus providing source portability of the functionality (if not runtime portability). At least this is how I've parsed @MrSidims's comment.

Also, it is often the case that users employing target-specific attributes like "FPGA pipelining" do not require their sources to be instantly portable from the chosen target and/or rely on the preprocessor to define different code paths for, say, FPGA and CPU. Overall, the concerns about some compiled device objects with target-tuned code not being executable on other targets don't seem to be severe for DPC++ users - unless, of course, I'm missing some data that shows otherwise.

**Comment by keryell** [comment]
At least just ignoring the FPGA features either in Clang or in the SYCL runtime is an easy.
I am claiming everywhere that a super cool feature of SYCL is that it allows to emulate FPGA code on GPU in a super-fast way compared to CPU or RTL simulation, so I would be sad this is not possible! :-)
What happens in the GPU SPIR-V consumer if the SPIR-V has FPGA extensions with no semantic changed could be allowed too.
But probably filtering them out earlier is clearer and some conditional warnings can be optionally emitted.

**Comment by MrSidims** [comment]
> I am claiming everywhere that a super cool feature of SYCL is that it allows to emulate FPGA code on GPU in a super-fast way compared to CPU or RTL simulation, so I would be sad this is not possible! :-)

(if only we could emulate DPCPP pipes on GPU (well, we actually could, but it's not working right now) :) )
We still can do that, most of these extensions are added to translate a certain LLVM metadata to SPIR-V. And whilst metadata can be easily ignored and removed from the IR module it's not always the case for SPIR-V: we indeed can safely remove decorations, but not instructions. So it would be a bit hard to support FPGA SPIR-V extensions in non-FPGA SPIR-V-centric devices' backends, and also not every FPGA attribute can be used to optimize IR for non-FPGA targets. The fact that if an extension is disabled in the driver doesn't mean, that the translator will emit an error - instead it will change translation flow and don't generate IR patterns appropriate for the extension.

There are also extensions (for now it's the only SPV_INTEL_usm_storage_classes) that can't be just ignored as it adds translation of not-common OpenCL address spaces. And it seems to be reasonable to deal with it in the 'middle-end' instead of different devices' backends.

**Review by mlychkov — APPROVED** [review]

**Review by AGindinson — APPROVED** [review]

**Pull request merged**